src/appleseed/renderer/modeling/bsdf/hairbsdf.cpp

```diff
@@ -0,0 +1,840 @@
+            if (p == 2) {
+                sin_theta_i = sample_longitudinal(
+                    std::sin(theta_o_TRT),
+                    std::cos(theta_o_TRT),
+                    values->m_precomputed.m_v * 4.f,
+                    sample_M);
+            }
+            float cos_theta_i = std::sqrt(std::max(0.0f, 1.f - sin_theta_i * sin_theta_i));
+
+
+            // Sample azimuthal scattering function to compute change in azimuthal direction
+            float dphi;
+            dphi = calc_phi(p, gamma_o, gamma_t) +
+            sample_trimmed_logistic(sample_N[1], values->m_precomputed.m_s);
+            
+            // Compute wi from sampled hair scattering angles
+            float phi_i = phi_o + dphi;
+            Vector3f wi = Vector3f(sin_theta_i, cos_theta_i * std::sin(phi_i), cos_theta_i * std::cos(phi_i));
+
+            Spectrum bsdf_f(0.f);
+            float bsdf_pdf(0.f);
+
+            // RR contribution of scattering
+            bsdf_f +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_R)),
+                    sin_theta_i, std::sin(theta_o_R), values->m_precomputed.m_v) *
+                ap[0] * azimuthal(dphi, 0, values->m_precomputed.m_s, gamma_o, gamma_t);
+            bsdf_pdf +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_R)),
+                    sin_theta_i, std::sin(theta_o_R), values->m_precomputed.m_v) *
+                ap_pdf[0] * azimuthal(dphi, 0, values->m_precomputed.m_s, gamma_o, gamma_t);
+
+            // TT contribution of scattering
+            bsdf_f +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_TT)), sin_theta_i,
+                    std::sin(theta_o_TT), values->m_precomputed.m_v * .25f) *
+                ap[1] * azimuthal(dphi, 1, values->m_precomputed.m_s, gamma_o, gamma_t);
+            bsdf_pdf +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_TT)), sin_theta_i,
+                    std::sin(theta_o_TT), values->m_precomputed.m_v * .25f) *
+                ap_pdf[1] * azimuthal(dphi, 1, values->m_precomputed.m_s, gamma_o, gamma_t);
+
+            // TRT contribution of scattering
+            bsdf_f +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_TRT)), sin_theta_i,
+                    std::sin(theta_o_TRT), values->m_precomputed.m_v * 4.f) *
+                ap[2] * azimuthal(dphi, 2, values->m_precomputed.m_s, gamma_o, gamma_t);
+            bsdf_pdf +=
+                longitudinal(cos_theta_i, std::abs(std::cos(theta_o_TRT)), sin_theta_i,
+                    std::sin(theta_o_TRT), values->m_precomputed.m_v * 4.f) *
+                ap_pdf[2] * azimuthal(dphi, 2, values->m_precomputed.m_s, gamma_o, gamma_t);
+
+            // TRRT+ contribution of scattering
+            bsdf_f +=
+                longitudinal(cos_theta_i, cos_theta_o, sin_theta_i, sin_theta_o, values->m_precomputed.m_v * 4.f) *
+                ap[3] / static_cast<float>(2.0f * pi);
```

Comment:
You can multiply by `RcpTwoPi<float>()` here (and everything else). Check `foundation/math/scalar.h` for the list of available constants.